README.rst
aiozipkin
=========
.. image:: https://travis-ci.com/aio-libs/aiozipkin.svg?branch=master
    :target: https://travis-ci.com/aio-libs/aiozipkin
.. image:: https://codecov.io/gh/aio-libs/aiozipkin/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/aio-libs/aiozipkin
.. image:: https://api.codeclimate.com/v1/badges/1ff813d5cad2d702cbf1/maintainability
   :target: https://codeclimate.com/github/aio-libs/aiozipkin/maintainability
   :alt: Maintainability
.. image:: https://img.shields.io/pypi/v/aiozipkin.svg
    :target: https://pypi.python.org/pypi/aiozipkin
.. image:: https://readthedocs.org/projects/aiozipkin/badge/?version=latest
    :target: http://aiozipkin.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status
.. image:: https://badges.gitter.im/Join%20Chat.svg
    :target: https://gitter.im/aio-libs/Lobby
    :alt: Chat on Gitter

**aiozipkin** is Python 3.6+ module that adds distributed tracing capabilities
from asyncio_ applications with zipkin (http://zipkin.io) server instrumentation.

zipkin_ is a distributed tracing system. It helps gather timing data needed
to troubleshoot latency problems in microservice architectures. It manages
both the collection and lookup of this data. Zipkin’s design is based on
the Google Dapper paper.

Applications are instrumented with  **aiozipkin** report timing data to zipkin_.
The Zipkin UI also presents a Dependency diagram showing how many traced
requests went through each application. If you are troubleshooting latency
problems or errors, you can filter or sort all traces based on the
application, length of trace, annotation, or timestamp.

.. image:: https://raw.githubusercontent.com/aio-libs/aiozipkin/master/docs/zipkin_animation2.gif
    :alt: zipkin ui animation


Features
========
* Distributed tracing capabilities to **asyncio** applications.
* Support zipkin_ ``v2`` protocol.
* Easy to use API.
* Explicit context handling, no thread local variables.
* Can work with jaeger_ and stackdriver_ through zipkin compatible API.


zipkin vocabulary
-----------------
Before code lets learn important zipkin_ vocabulary, for more detailed
information please visit https://zipkin.io/pages/instrumenting

.. image:: https://raw.githubusercontent.com/aio-libs/aiozipkin/master/docs/zipkin_glossary.png
    :alt: zipkin ui glossary

* **Span** represents one specific method (RPC) call
* **Annotation** string data associated with a particular timestamp in span
* **Tag** - key and value associated with given span
* **Trace** - collection of spans, related to serving particular request


Simple example
--------------

.. code:: python

    import asyncio
    import aiozipkin as az


    async def run():
        # setup zipkin client
        zipkin_address = 'http://127.0.0.1:9411/api/v2/spans'
        endpoint = az.create_endpoint(
            "simple_service", ipv4="127.0.0.1", port=8080)
        tracer = await az.create(zipkin_address, endpoint, sample_rate=1.0)

        # create and setup new trace
        with tracer.new_trace(sampled=True) as span:
            # give a name for the span
            span.name("Slow SQL")
            # tag with relevant information
            span.tag("span_type", "root")
            # indicate that this is client span
            span.kind(az.CLIENT)
            # make timestamp and name it with START SQL query
            span.annotate("START SQL SELECT * FROM")
            # imitate long SQL query
            await asyncio.sleep(0.1)
            # make other timestamp and name it "END SQL"
            span.annotate("END SQL")

        await tracer.close()

    if __name__ == "__main__":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(run())


aiohttp example
---------------

*aiozipkin* includes *aiohttp* server instrumentation, for this create
`web.Application()` as usual and install aiozipkin plugin:


.. code:: python

    import aiozipkin as az

    def init_app():
        host, port = "127.0.0.1", 8080
        app = web.Application()
        endpoint = az.create_endpoint("AIOHTTP_SERVER", ipv4=host, port=port)
        tracer = await az.create(zipkin_address, endpoint, sample_rate=1.0)
        az.setup(app, tracer)


That is it, plugin adds middleware that tries to fetch context from headers,
and create/join new trace. Optionally on client side you can add propagation
headers in order to force tracing and to see network latency between client and
server.

.. code:: python

    import aiozipkin as az

    endpoint = az.create_endpoint("AIOHTTP_CLIENT")
    tracer = await az.create(zipkin_address, endpoint)

    with tracer.new_trace() as span:
        span.kind(az.CLIENT)
        headers = span.context.make_headers()
        host = "http://127.0.0.1:8080/api/v1/posts/{}".format(i)
        resp = await session.get(host, headers=headers)
        await resp.text()


Documentation
-------------
http://aiozipkin.readthedocs.io/


Installation
------------
Installation process is simple, just::

    $ pip install aiozipkin


Support of other collectors
===========================
**aiozipkin** can work with any other zipkin_ compatible service, currently we
tested it with jaeger_ and stackdriver_.


Jaeger support
--------------
jaeger_ supports zipkin_ span format as result it is possible to use *aiozipkin*
with jaeger_ server. You just need to specify *jaeger* server address and it
should work out of the box. Not need to run local zipkin server.
For more informations see tests and jaeger_ documentation.

.. image:: https://raw.githubusercontent.com/aio-libs/aiozipkin/master/docs/jaeger.png
    :alt: jaeger ui animation


Stackdriver support
-------------------
Google stackdriver_ supports zipkin_ span format as result it is possible to
use *aiozipkin* with this google_ service. In order to make this work you
need to setup zipkin service locally, that will send trace to the cloud. See
google_ cloud documentation how to setup make zipkin collector:

.. image:: https://raw.githubusercontent.com/aio-libs/aiozipkin/master/docs/stackdriver.png
    :alt: jaeger ui animation


Requirements
------------

* Python_ 3.6+
* aiohttp_


.. _PEP492: https://www.python.org/dev/peps/pep-0492/
.. _Python: https://www.python.org
.. _aiohttp: https://github.com/KeepSafe/aiohttp
.. _asyncio: http://docs.python.org/3.5/library/asyncio.html
.. _uvloop: https://github.com/MagicStack/uvloop
.. _zipkin: http://zipkin.io
.. _jaeger: http://jaeger.readthedocs.io/en/latest/
.. _stackdriver: https://cloud.google.com/stackdriver/
.. _google: https://cloud.google.com/trace/docs/zipkin


CHANGES.rst
CHANGES
=======

..
    You should *NOT* be adding new change log entries to this file, this
    file is managed by towncrier. You *may* edit previous change logs to
    fix problems like typo corrections or such.
    To add a new change log entry, please see
    https://pip.pypa.io/en/latest/development/#adding-a-news-entry
    we named the news folder "changes".

    WARNING: Don't drop the next directive!

.. towncrier release notes start

1.1.1 (2021-10-23)
==================

Bugfixes
--------

- Fix unhandled AssertionError in aiohttp integration when unknown resource requested by the user.
  `#400 <https://github.com/aio-libs/aiozipkin/issues/400>`_
- Fix ``NoneType`` error when using ``SystemRoute``.
  `#410 <https://github.com/aio-libs/aiozipkin/issues/410>`_

----


1.1.0 (2021-05-17)
==================

Bugfixes
--------

- Expect trace request context to be of SimpleNamespace type.
  `#385 <https://github.com/aio-libs/aiozipkin/issues/385>`_


----


1.0.0 (2020-11-06)
==================

Bugfixes
--------

- Support Python 3.8 and Python 3.9
  `#259 <https://github.com/aio-libs/aiozipkin/issues/259>`_


----


0.7.1 (2020-09-20)
==================

Bugfixes
--------

- Fix `Manifest.in` file; add `CHANGES.rst` to the Source Tarball.


0.7.0 (2020-07-17)
==================

Features
--------

- Add support of AWS X-Ray trace id format.
  `#273 <https://github.com/aio-libs/aiozipkin/issues/273>`_


----


0.6.0 (2019-10-12)
------------------
* Add context var support for python3.7 aiohttp instrumentation #187
* Single header tracing support #189
* Add retries and batches to transport (thanks @konstantin-stepanov)
* Drop python3.5 support #238
* Use new typing syntax in codebase #237


0.5.0 (2018-12-25)
------------------
* More strict typing configuration is used #147
* Fixed bunch of typos in code and docs #151 #153 (thanks @deejay1)
* Added interface for Transport #155 (thanks @deejay1)
* Added create_custom helper for easer tracer configuration #160 (thanks @deejay1)
* Added interface for Sampler #160 (thanks @deejay1)
* Added py.typed marker


0.4.0 (2018-07-11)
------------------
* Add more coverage with typing #147
* Breaking change: typo send_inteval => send_interval #144 (thanks @gugu)
* Breaking change: do not append api/v2/spans to the zipkin dress #150


0.3.0 (2018-06-13)
------------------
* Add support http.route tag for aiohttp #138
* Make zipkin address builder more permissive #141 (thanks @dsantosfff)


0.2.0 (2018-03-03)
------------------
* Breaking change: az.create is coroutine now #114
* Added context manger for tracer object #114
* Added more mypy types #117


0.1.1 (2018-01-26)
------------------
* Added new_child helper method #83


0.1.0 (2018-01-21)
------------------
After few months of work and beta releases here are basic features:

* Initial release.
* Implemented zipkin v2 protocol with HTTP transport
* Added jaeger support
* Added stackdriver support
* Added aiohttp server support
* Added aiohttp 3.0.0 client tracing support
* Added examples and demos


examples/simple.py
import asyncio

import aiozipkin as az


async def run() -> None:
    # setup zipkin client
    zipkin_address = "http://127.0.0.1:9411/api/v2/spans"
    endpoint = az.create_endpoint("simple_service", ipv4="127.0.0.1", port=8080)

    # creates tracer object that traces all calls, if you want sample
    # only 50% just set sample_rate=0.5
    tracer = await az.create(zipkin_address, endpoint, sample_rate=1.0)

    # create and setup new trace
    with tracer.new_trace(sampled=True) as span:
        span.name("root_span")
        span.tag("span_type", "root")
        span.kind(az.CLIENT)
        span.annotate("SELECT * FROM")
        # imitate long SQL query
        await asyncio.sleep(0.1)
        span.annotate("start end sql")

        # create child span
        with tracer.new_child(span.context) as nested_span:
            nested_span.name("nested_span_1")
            nested_span.kind(az.CLIENT)
            nested_span.tag("span_type", "inner1")
            nested_span.remote_endpoint("remote_service_1")
            await asyncio.sleep(0.01)

        # create other child span
        with tracer.new_child(span.context) as nested_span:
            nested_span.name("nested_span_2")
            nested_span.kind(az.CLIENT)
            nested_span.remote_endpoint("remote_service_2")
            nested_span.tag("span_type", "inner2")
            await asyncio.sleep(0.01)

    await tracer.close()
    print("-" * 100)
    print("Check zipkin UI for produced traces: http://localhost:9411/zipkin")


if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(run())
